File: python/second/sliding_window_maximum.py
class Solution(object):
    # O(n) time | O(n) space
    def maxSlidingWindow(self, nums, k):
        """
        :type nums: List[int]
        :type k: int
        :rtype: List[int]
        """
        queue = []
        res = []

        for i in range(len(nums)):
            num = nums[i]

            if len(queue) > 0 and i - k == queue[0][1]:
                queue.pop(0) # remove fromt

            while len(queue) > 0 and queue[-1][0] < num:
                queue.pop() # remove back
            
            queue.append([num, i]) # add back
            if i >= k - 1:
                res.append(queue[0][0])
        
        return res

solution = Solution()
nums = [1,3,-1,-3,5,3,6,7]
k = 3
# nums = [1,3,1,2,0,5]
# k = 3
print(solution.maxSlidingWindow(nums, k))

File: src/main/java/september_2025/SlidingWindowMaximum.java
package september_2025;

import java.util.ArrayDeque;

public class SlidingWindowMaximum {

    public static void main(String[] args) {
        int[] nums = {1,3,-1,-3,2,3,6,7};
        int k = 1;

        SlidingWindowMaximum slidingWindowMaximum = new SlidingWindowMaximum();
        int[] result = slidingWindowMaximum.maxSlidingWindow(nums, k);
        System.out.println(result);
    }

    // O(n) time | O(n) space
    public int[] maxSlidingWindow(int[] nums, int k) {
        int[] res = new int[nums.length - k + 1];
        int idx = 0;
        ArrayDeque<int[]> queue = new ArrayDeque<>(); // <num, idx>

        for (int i = 0; i < nums.length; i++) {
            // add to end of the queue
            while (!queue.isEmpty() && nums[i] > queue.getLast()[0]) {
                queue.removeLast();
            }
            queue.addLast(new int[] {nums[i], i});
            // check whether we need to remove element from front of queue
            int firstIdx = queue.getFirst()[1];
            if (i - k == firstIdx) {
                queue.removeFirst();
            }
            if (i >= k - 1) {
                int[] first = queue.getFirst();
                res[idx] = first[0];
                idx++;
            }
        }
        for (int i = idx; i < res.length; i++) {
            res[i] = queue.getFirst()[0];
        }
        return res;
    }

}
